File: problem-003/solution.py
from typing import List


class Node:
    """
    Node is a binary tree node.
    """

    def __init__(self, val: str, left: Node = None, right: Node = None):
        self.val = val
        self.left = left
        self.right = right


END_MARKER = "end"


def serialize(root: Node) -> str:
    """
    Convert a root of a binary tree into a string representation.
    Just recursively call this function with inorder traversal.
    Each node will be separated by comma.
    """
    if root is None:
        return END_MARKER

    left = serialize(root.left)
    right = serialize(root.right)
    return f"{root.val},{left},{right}"


def deserialize(tree_string: str) -> Node:
    """
    Return the root node of a binary tree given its string representation.
    Split the string representation into list of strings, reverse it, and pass it to helper function.
    """
    values = tree_string.split(",")[::-1]
    return _deserialize(values)


def _deserialize(values: List[str]) -> Node:
    """
    1. Pop last element from the list, create root node.
    2. Generate left and right node recursively.
    """
    if not values:
        return None

    value = values.pop()
    if value == END_MARKER:
        return None

    root = Node(value)
    root.left = _deserialize(values)
    root.right = _deserialize(values)
    return root


node = Node('root', Node('left', Node('left.left')), Node('right'))
assert deserialize(serialize(node)).left.left.val == 'left.left'

File: problem-003/solution.go
package problem003

import (
	"fmt"
	"strings"
)

type treeNode struct {
	val   string
	left  *treeNode
	right *treeNode
}

// serialize takes a root node and serializes it into a string
// represenration with inorder traversal.
// Each node will be separated by comma, and null represented with "end".
func serialize(root *treeNode) string {
	if root == nil {
		return "end"
	}

	left := serialize(root.left)
	right := serialize(root.right)
	return fmt.Sprintf("%s,%s,%s", root.val, left, right)
}

// deserialize takes a string representation of a binary tree,
// converts it back into a binary tree, and returns the root node.
func deserialize(s string) *treeNode {
	if len(s) == 0 {
		return nil
	}

	nodeVals := strings.Split(s, ",")
	return deserializeHelper(nodeVals)
}

// deserializeHelper takes a slice of node values of a binary tree,
// converts it back into a binary tree and returns the root node.
func deserializeHelper(nodeVals []string) *treeNode {
	if len(nodeVals) == 0 {
		return nil
	}

	val := nodeVals[0]
	copy(nodeVals, nodeVals[1:])

	if val == "end" {
		return nil
	}

	root := &treeNode{
		val:   val,
		left:  deserializeHelper(nodeVals),
		right: deserializeHelper(nodeVals),
	}

	return root
}
